### ISSUE TITLE
Multiple calls to linksCache.expire/fetchHighlights on startup in a new profile

### ISSUE THREAD

**Issue by piatra** [mozilla/activity-stream#3724]
> One thing I noticed when testing the PR, but this probably should be its own issue/PR. #3675 added expiring the highlights links cache on BOOKMARK_ADDED. A new profile causes adds of several bookmarks causing the cache to be expired and fetched from places multiple times on startup.

from https://github.com/mozilla/activity-stream/pull/3695#issuecomment-337165574

**Comment by piatra** [comment]
We need to invalidate the cache otherwise we won't get newly bookmarked items into highlights until the cache expires on its own.
 * I thought about using the `action.data` field but that would insert the default bookmarks

The alternative to that which I tried and seems to work is to use a setTimeout ~500ms which we can clear on consecutive calls (like startup).
 * not sure if there are other timing issues this might cause

Anything else we could try? @Mardak

**Comment by Mardak** [comment]
When we get the computation for profile age cutoff, we could skip PLACES_BOOKMARK_ADDED if that time has not expired.

**Comment by piatra** [comment]
Do you mean something like 
```
if Date.now() < ProfileAge + CUTOFF
  return;
```

At this time I don't want to go forward with the profile age cutoff time because:
 * on old hardware that number is significant (32s in my tests) and could interfere with importing a browsing profile (on a fast machine).
 * ProfileAge caused all those failures on try
 * when you create an additional profile (from the default one) the bookmarks are not automatically added until you actually start it (which are different actions) and no amount of cutoff could account for that.

For that issue in particular I was looking towards other solutions:
 * querying for bookmarks on startup and saving in a pref https://github.com/mozilla/activity-stream/blob/testpilot-legacy/addon/PlacesProvider.js#L796-L808 
 * looking into a more complex bookmarks query https://searchfox.org/mozilla-central/source/toolkit/components/places/Bookmarks.jsm#1699

Both of which won't probably help this issue.

**Comment by Mardak** [comment]
The approach doesn't need to use `ProfileAge`, but conceptually, if we're having the query filter out bookmarks based on time and we see a bookmark just added during that "should be filtered out" time, we shouldn't need to fetch because it would be removed anyway.

**Comment by Mardak** [comment]
Alternatively, if the behavior we want is "don't show unvisited bookmarks", we could check if the jsut added bookmark has visits before expiring the cache and fetching.

**Comment by piatra** [comment]
> Alternatively, if the behavior we want is "don't show unvisited bookmarks", we could check if the just added bookmark has visits before expiring the cache and fetching.

I don't think we can rely on this if we want bookmarks from Top Stories to show up.

**Comment by Mardak** [comment]
So it sounds like this depends quite a bit on the approach to fixing #3131. One possible fix for that issue is to add a visit to the story url before bookmarking it.

**Comment by piatra** [comment]
Maybe it doesn't depend on #3131 (and the solution for that issue isn't time based filtering anymore it's just adding the extra visit). Now that #3695 landed and INIT always happens first we could save the time when it occurred and and skip `cache.refresh` calls for BOOKMARK_ADDED events for the first 5-10s. Migration is already a separate event that has `broadcast: true`.
Would this solution be too specific for this issue?

**Comment by Mardak** [comment]
Have you looked at what data is available when something is bookmarked? Is there any data there that would indicate we shouldn't `expire` or not even bother dispatching PLACES_BOOKMARK_ADDED?

**Comment by piatra** [comment]
@Mardak great idea. Looks like there are 2 cases for the default bookmarks:
 * All valid URLs bookmarked (meaning they are not not about:* pages or folders) [have a source value of 3](https://searchfox.org/mozilla-central/source/toolkit/components/places/nsINavBookmarksService.idl#322)
 * about:* pages and folders have a source value of 0 but they can easily be skipped if we `try` to create a new URL instance with that value

All regular bookmarks you add have a source value of 0.
We can check the source before [dispatching PLACES_BOOKMARK_ADDED](https://github.com/mozilla/activity-stream/blob/master/system-addon/lib/PlacesFeed.jsm#L104) events.

**Comment by Mardak** [comment]
Sounds reasonable to short circuit similar to the existing `type !== PlacesUtils.bookmarks.TYPE_BOOKMARK` check but now additionally for source value 3 and non-http/s urls. #3722 would need a fix in NewTabUtils anyway to match non-http/s: https://searchfox.org/mozilla-central/source/toolkit/modules/NewTabUtils.jsm#860-869
